### ISSUE TITLE
Bad file descriptor when building multi-module project

### ISSUE THREAD

**Issue by velo** [spotify/docker-maven-plugin#115]
Hi,

I have a project with multiple modules and when I build it using multiple threads (-T2) I get the following error:

```
[ERROR] Failed to execute goal com.spotify:docker-maven-plugin:0.2.13-SNAPSHOT:build (default) on project company-project: Exception caught: java.io.IOException: Bad file descriptor -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.spotify:docker-maven-plugin:0.2.13-SNAPSHOT:build (default) on project company-project: Exception caught
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:216)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:189)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:185)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.maven.plugin.MojoExecutionException: Exception caught
    at com.spotify.docker.AbstractDockerMojo.execute(AbstractDockerMojo.java:116)
    at com.spotify.docker.BuildMojo.execute(BuildMojo.java:78)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
    ... 11 more
Caused by: com.spotify.docker.client.DockerException: java.io.IOException: Bad file descriptor
    at com.spotify.docker.client.ProgressStream.hasNextMessage(ProgressStream.java:57)
    at com.spotify.docker.client.DefaultDockerClient.build(DefaultDockerClient.java:818)
    at com.spotify.docker.BuildMojo.buildImage(BuildMojo.java:465)
    at com.spotify.docker.BuildMojo.execute(BuildMojo.java:270)
    at com.spotify.docker.AbstractDockerMojo.execute(AbstractDockerMojo.java:114)
    ... 14 more
Caused by: java.io.IOException: Bad file descriptor
    at jnr.enxio.channels.NativeSocketChannel.read(NativeSocketChannel.java:82)
    at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:59)
    at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:109)
    at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103)
    at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:136)
    at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:152)
    at org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:270)
    at org.apache.http.impl.io.AbstractMessageParser.parseHeaders(AbstractMessageParser.java:192)
    at org.apache.http.impl.io.AbstractMessageParser.parseHeaders(AbstractMessageParser.java:145)
    at org.apache.http.impl.io.ChunkedInputStream.parseTrailerHeaders(ChunkedInputStream.java:264)
    at org.apache.http.impl.io.ChunkedInputStream.nextChunk(ChunkedInputStream.java:214)
    at org.apache.http.impl.io.ChunkedInputStream.read(ChunkedInputStream.java:169)
    at org.apache.http.conn.EofSensorInputStream.read(EofSensorInputStream.java:137)
    at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
    at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
    at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
    at java.io.FilterInputStream.read(FilterInputStream.java:133)
    at org.glassfish.jersey.message.internal.EntityInputStream.read(EntityInputStream.java:101)
    at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream.read(ReaderInterceptorExecutor.java:300)
    at com.spotify.docker.client.shaded.com.fasterxml.jackson.core.json.UTF8StreamJsonParser.loadMore(UTF8StreamJsonParser.java:169)
    at com.spotify.docker.client.shaded.com.fasterxml.jackson.core.json.UTF8StreamJsonParser._skipWSOrEnd(UTF8StreamJsonParser.java:2495)
    at com.spotify.docker.client.shaded.com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:617)
    at com.spotify.docker.client.shaded.com.fasterxml.jackson.databind.MappingIterator.hasNextValue(MappingIterator.java:159)
    at com.spotify.docker.client.ProgressStream.hasNextMessage(ProgressStream.java:53)
    ... 18 more

```

I created a test that simulates the problem at:
https://github.com/velo/docker-maven-plugin/commit/a599b7b63ed006f0a4c8c89975f54ea25f7815bb
(note it uses maven 3.3.3 maven.config to set -T10 https://github.com/velo/docker-maven-plugin/commit/a599b7b63ed006f0a4c8c89975f54ea25f7815bb#diff-d235c7fd6aa195ea001067a1e7113ce5R1)

I looked around and seems something that blows inside docker.  Since at the moment I can not investigate it further, I can up with a lock to prevent 2 docker images being built at the exact same time:
https://github.com/velo/docker-maven-plugin/commit/cb099b604861fd327e1a329ef590474edc17c0ab

If that is acceptable and you guys would like to incorporate to this project:
https://github.com/spotify/docker-maven-plugin/pull/114

**Comment by davidxia** [comment]
@velo Thanks for your contribution. I ran `mvn clean verify` on just the ["Testing docker-maven-plugin using multiple threads" commit here][1] and expected it to fail, but my tests passed without the lock.

I'm using boot2docker 1.6.2 on a Mac with Maven 3.3.3. Any ideas why it works for me?

```
[INFO] Building: parallel-modules/pom.xml
[INFO] [INFO] Scanning for projects...
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] Reactor Build Order:
[INFO] [INFO]
[INFO] [INFO] parallel-modules
[INFO] [INFO] parallel-modules-a
[INFO] [INFO] parallel-modules-b
[INFO] [INFO]
[INFO] [INFO] Using the MultiThreadedBuilder implementation with a thread count of 10
[INFO] [INFO]
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] Building parallel-modules 1.0-SNAPSHOT
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO]
[INFO] [INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ parallel-modules ---
[INFO] [INFO]
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO]
[INFO] [INFO] Building parallel-modules-a 1.0-SNAPSHOT
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] Building parallel-modules-b 1.0-SNAPSHOT
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO]
[INFO] [INFO]
[INFO] [INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ parallel-modules-b ---
[INFO] [INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ parallel-modules-a ---
[INFO] [INFO]
[INFO] [INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ parallel-modules-a ---
[INFO] [INFO]
[INFO] [INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ parallel-modules-b ---
[INFO] [INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] [INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] [INFO] skip non existing resourceDirectory /Users/david/src/docker-maven-plugin/target/it/parallel-modules/b/src/main/resources
[INFO] [INFO] skip non existing resourceDirectory /Users/david/src/docker-maven-plugin/target/it/parallel-modules/a/src/main/resources
[INFO] [INFO]
[INFO] [INFO] --- maven-compiler-plugin:2.5.1:compile (default-compile) @ parallel-modules-b ---
[INFO] [INFO]
[INFO] [INFO] --- maven-compiler-plugin:2.5.1:compile (default-compile) @ parallel-modules-a ---
[INFO] [INFO] Compiling 1 source file to /Users/david/src/docker-maven-plugin/target/it/parallel-modules/a/target/classes
[INFO] [INFO] Compiling 1 source file to /Users/david/src/docker-maven-plugin/target/it/parallel-modules/b/target/classes
[INFO] [INFO]
[INFO] [INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ parallel-modules-a ---
[INFO] [INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] [INFO] skip non existing resourceDirectory /Users/david/src/docker-maven-plugin/target/it/parallel-modules/a/src/test/resources
[INFO] [INFO]
[INFO] [INFO] --- maven-compiler-plugin:2.5.1:testCompile (default-testCompile) @ parallel-modules-a ---
[INFO] [INFO] No sources to compile
[INFO] [INFO]
[INFO] [INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ parallel-modules-a ---
[INFO] [INFO]
[INFO] [INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ parallel-modules-b ---
[INFO] [INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] [INFO] skip non existing resourceDirectory /Users/david/src/docker-maven-plugin/target/it/parallel-modules/b/src/test/resources
[INFO] [INFO]
[INFO] [INFO] --- maven-compiler-plugin:2.5.1:testCompile (default-testCompile) @ parallel-modules-b ---
[INFO] [INFO] No sources to compile
[INFO] [INFO]
[INFO] [INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ parallel-modules-b ---
[INFO] [INFO] No tests to run.
[INFO] [INFO] No tests to run.
[INFO] [INFO]
[INFO] [INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ parallel-modules-b ---
[INFO] [INFO]
[INFO] [INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ parallel-modules-a ---
[INFO] [INFO] Building jar: /Users/david/src/docker-maven-plugin/target/it/parallel-modules/b/target/parallel-modules-b-1.0-SNAPSHOT.jar
[INFO] [INFO] Building jar: /Users/david/src/docker-maven-plugin/target/it/parallel-modules/a/target/parallel-modules-a-1.0-SNAPSHOT.jar
[INFO] [INFO]
[INFO] [INFO] --- docker-maven-plugin:0.2.13-SNAPSHOT:build (default-build) @ parallel-modules-a ---
[INFO] [INFO]
[INFO] [INFO] --- docker-maven-plugin:0.2.13-SNAPSHOT:build (default-build) @ parallel-modules-b ---
[INFO] [INFO] Copying /Users/david/src/docker-maven-plugin/target/it/parallel-modules/a/target/parallel-modules-a-1.0-SNAPSHOT.jar -> /Users/david/src/docker-maven-plugin/target/it/parallel-modules/a/target/docker/parallel-modules-a-1.0-SNAPSHOT.jar
[INFO] [INFO] Copying /Users/david/src/docker-maven-plugin/target/it/parallel-modules/b/target/parallel-modules-b-1.0-SNAPSHOT.jar -> /Users/david/src/docker-maven-plugin/target/it/parallel-modules/b/target/docker/parallel-modules-b-1.0-SNAPSHOT.jar
[INFO] [INFO] Building image parallel-modules-a
[INFO] [INFO] Building image parallel-modules-b
[INFO] Step 0 : FROM java
[INFO] Step 0 : FROM java
[INFO] ---> beabd1eef902
[INFO] ---> beabd1eef902
[INFO] Step 1 : ENTRYPOINT java -jar /parallel-modules-a-1.0-SNAPSHOT.jar
[INFO] Step 1 : ENTRYPOINT java -jar /parallel-modules-b-1.0-SNAPSHOT.jar
[INFO] ---> Using cache
[INFO] ---> 42ee3bc2bec9
[INFO] Step 2 : CMD
[INFO] ---> Using cache
[INFO] ---> 0971cbde4fe7
[INFO] Step 2 : CMD
[INFO] ---> Using cache
[INFO] ---> 8aa1c6bb5933
[INFO] Step 3 : ADD /parallel-modules-b-1.0-SNAPSHOT.jar /parallel-modules-b-1.0-SNAPSHOT.jar
[INFO] ---> Using cache
[INFO] ---> 5e284495e310
[INFO] Step 3 : ADD /parallel-modules-a-1.0-SNAPSHOT.jar /parallel-modules-a-1.0-SNAPSHOT.jar
[INFO] ---> 77911b5e8bea
[INFO] ---> d4f1d3429f1d
[INFO] Removing intermediate container 7c31e1fb9583
[INFO] Successfully built 77911b5e8bea
[INFO] [INFO] Built parallel-modules-a
[INFO] Removing intermediate container 0940553382a4
[INFO] Successfully built d4f1d3429f1d
[INFO] [INFO] Built parallel-modules-b
[INFO] [INFO]
[INFO] [INFO] --- maven-enforcer-plugin:1.3.1:enforce (enforce-files-exist) @ parallel-modules-a ---
[INFO] [INFO]
[INFO] [INFO] --- maven-enforcer-plugin:1.3.1:enforce (enforce-files-exist) @ parallel-modules-b ---
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] Reactor Summary:
[INFO] [INFO]
[INFO] [INFO] parallel-modules ................................... SUCCESS [  0.095 s]
[INFO] [INFO] parallel-modules-a ................................. SUCCESS [  6.513 s]
[INFO] [INFO] parallel-modules-b ................................. SUCCESS [  6.513 s]
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] BUILD SUCCESS
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] Total time: 8.813 s (Wall Clock)
[INFO] [INFO] Finished at: 2015-08-02T22:48:02-04:00
[INFO] [INFO] Final Memory: 26M/236M
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] ..SUCCESS (11.8 s)
```

  [1]: https://github.com/velo/docker-maven-plugin/commit/a599b7b63ed006f0a4c8c89975f54ea25f7815bb

**Comment by davidxia** [comment]
also tried on ubuntu 14.04 with docker 1.6.2

**Comment by velo** [comment]
hi @davidxia 
I'm using ubuntu 1504 and docker 1.7.1

no boot2docker for me as well

**Comment by davidxia** [comment]
@velo I ran `mvn clean verify` on the ["Testing docker-maven-plugin using multiple threads" commit here][1] on ubuntu 14.04 and docker 1.7.1 but the test still passed. I'm wondering how I can repro your error.

**Comment by velo** [comment]
Is there any other information from my environment that I can share to help figure out this?

**Comment by dawidmalina** [comment]
Today one of my colleagues showed me similar stacktrace. I will try to check with him is this issue is related to -T option. If yes then I will try to use this patch.

**Comment by velo** [comment]
Outstanding @dawidmalina 
=)

**Comment by dawidmalina** [comment]
Heh today I can't reproduce this issue even with -T30 :( and yes this is issue related to docker not plugin.

**Comment by davidxia** [comment]
In that case can I close?

On Friday, August 7, 2015, Dawid Malinowski <notifications@github.com>
wrote:

> Heh today I can't reproduce this issue even with -T30 :( and yes this is
> issue related to docker not plugin.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-maven-plugin/issues/115#issuecomment-128655680>
> .
>


-- 
David Xia
Software Engineer
dxia@spotify.com

**Comment by velo** [comment]
@dawidmalina 

What did you do that solved your issue?

I still have it.

May be, we could include this with a configuration?

**Comment by dawidmalina** [comment]
I didn't do anything. We had this issue only one day. Yes I know it is strange.

**Comment by velo** [comment]
=/
=(
=\

**Comment by davidxia** [comment]
should be fixed by #383
